modules/db/src/main/scala/gem/dao/StepDao.scala

```diff
@@ -362,4 +362,34 @@ object StepDao {
     } yield res
   }
 
+  /** Deletes all steps for the given observation, if any.
+    *
+    * @param oid observation whose steps should be deleted
+    */
+  def delete(oid: Observation.Id): ConnectionIO[Unit] = {
+    // Cascading delete takes care of the subtype steps like step_bias,
+    // step_dark, etc., but will leave a step_gcal's calibration configuration
+    // in the gcal table.  That means we have to explicitly remove the gcal
+    // configuration if we're deleting a gcal step.
+
+    val delGcal =
+      sql"""
+        DELETE FROM gcal g
+              USING step_gcal sg, step s
+              WHERE s.observation_id = $oid
+                AND s.step_id        = sg.step_gcal_id
+                AND sg.gcal_id       = g.gcal_id
+      """.update.run
```

Comment:
I certainly may have set it up incorrectly.  There is a `gcal` table which has gcal configurations with unique ids. Each gcal configuration corresponds to either a step in `step_gcal` or else a smart gcal mapping like those in `smart_f2`.  The reason I made a `gcal` table was to avoid duplicating all the gcal config columns in both `step_gcal` and every `smart_*` table like `smart_f2`.

So the `gcal` entries don't reference steps or smart gcal mappings but steps and smart gcal mappings reference them.  To make it cascade I'd need to add an optional step id reference in the `gcal` table?